test/functional/p2p_getaddr_caching.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test addr response caching
+"""
+
+from test_framework.messages import (
+    CAddress,
+    NODE_NETWORK,
+    NODE_WITNESS,
+    msg_addr,
+    msg_getaddr,
+)
+from test_framework.mininode import (
+    P2PInterface,
+    mininode_lock
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    wait_until
+)
+import time
+
+ADDRS = []
+for i in range(1000):
```

Comment:
Since all addresses go into the same bucket, I think that only 64 of them will be kept in addrman, 14 of which will be returned in response to each `GETADDR` - so creating 1000 addresses instead of 64 does not add randomness to the `GETADDR` responses - though it's fine of course if you want to test the 1000 address limit for `ADDR`.

Reply:
nice catch! Let's keep it this way just in case e.g. bucketing logic is changed in the future.

Reply:
Does it make sense to start off the node with > 1000 addresses? If I understand this test correctly, after the cache timeout the only thing that changes is the sequence of the same 1000 addresses in the message. I'm not sure if its worth it but we could also ensure that different addresses were sent after the timeout.

Reply:
I think that @mzumsande is right. Because these go into the same bucket, only a small number of them will be kept in addrman, and only 23% of those will be returned by `CAddrMan::GetAddr()`. I haven't verified the numbers 64 and 14, but that sounds right to me.

I suspect if we changed this to load >4000 addresses into addrman, then the test would fail, because responses to `GETADDR` messages would be different (see my comments about how the cache interacts with `vAddrToSend`. We should make that change, and also fix the code so it does what we're expecting!